### PR TITLE
feat: add kimi-vl transformers vision bot

### PR DESCRIPTION
**Pull request by weedge**
feat:
- add kimi-vl vision bot
- add kimi-vl transformers test
```shell
LLM_DEVICE=cuda LLM_MODEL_NAME_OR_PATH=./models/moonshotai/Kimi-VL-A3B-Instruct \
    python -m unittest test.core.llm.test_transformers_v_kimi.TestTransformersVKimi.test_chat_completion_prompts

LLM_DEVICE=cuda LLM_MODEL_NAME_OR_PATH=./models/moonshotai/Kimi-VL-A3B-Thinking \
    python -m unittest test.core.llm.test_transformers_v_kimi.TestTransformersVKimi.test_chat_completion_prompts
```
- kimi use qwen vision prompt
- deploy kimi-vl vision bot to modal
- add qwen2.5-vl vision bot
- add default splits layers into different GPUs when use device_map (use L4x2) ( layers.0是MLA+MLP 具体见：https://github.com/ai-bot-pro/achatbot/pull/142#issuecomment-2794506160 ）
```
{
    'language_model.model.layers.0': 0,
    'language_model.model.layers.1': 0,
    'language_model.model.layers.2': 0,
    'language_model.model.layers.3': 0,
    'language_model.model.layers.4': 0,
    'language_model.model.layers.5': 0,
    'language_model.model.layers.6': 0,
    'language_model.model.layers.7': 0,
    'language_model.model.layers.8': 0,
    'language_model.model.layers.9': 0,
    'language_model.model.layers.10': 0,
    'language_model.model.layers.11': 0,
    'language_model.model.layers.12': 0,
    'language_model.model.layers.13': 1,
    'language_model.model.layers.14': 1,
    'language_model.model.layers.15': 1,
    'language_model.model.layers.16': 1,
    'language_model.model.layers.17': 1,
    'language_model.model.layers.18': 1,
    'language_model.model.layers.19': 1,
    'language_model.model.layers.20': 1,
    'language_model.model.layers.21': 1,
    'language_model.model.layers.22': 1,
    'language_model.model.layers.23': 1,
    'language_model.model.layers.24': 1,
    'language_model.model.layers.25': 1,
    'language_model.model.layers.26': 1,
    'vision_tower': 0,
    'multi_modal_projector': 0,
    'language_model.model.embed_tokens': 0,
    'language_model.model.norm': 0,
    'language_model.lm_head': 0
}
```

fix:
- vision llm generate add  torch.cuda.empty_cache()
- torch.multiprocessing.set_start_method(spawn, force=True) use spawn run subprocessor need re-init log

---

A100 colab 笔记: 
- kimi-vl: https://github.com/weedge/doraemon-nb/blob/main/kimi_VL_A3B.ipynb
- achatbot+kimi-vl: https://github.com/weedge/doraemon-nb/blob/main/achatbot_kimivl.ipynb

config/bots/livekit_describe_transformers_vision_bot.json
```json
{
  "chat_bot_name": "LivekitDescribeVisionBot",
  "room_name": "chat-bot",
  "room_url": "",
  "token": "",
  "room_manager": {
    "tag": "livekit_room",
    "args": {
      "bot_name": "LivekitDescribeVisionBot",
      "is_common_session": false
    }
  },
  "services": {
    "pipeline": "achatbot",
    "vad": "silero",
    "asr": "sense_voice",
    "llm": "transformers_manual_vision_kimi",
    "tts": "edge"
  },
  "config": {
    "vad": {
      "tag": "silero_vad_analyzer",
      "args": { "stop_secs": 0.7 }
    },
    "asr": {
      "tag": "sense_voice_asr",
      "args": {
        "language": "zn",
        "model_name_or_path": "./models/FunAudioLLM/SenseVoiceSmall"
      }
    },
    "llm": {
      "tag": "llm_transformers_manual_vision_kimi",
      "args": {
        "lm_device": "cuda",
        "lm_model_name_or_path": "./models/moonshotai/Kimi-VL-A3B-Instruct",
        "chat_history_size": 0,
        "lm_torch_dtype":"bfloat16",
        "lm_attn_impl":"flash_attention_2",
        "model_type": "chat_completion"
      },
      "language": "zh"
    },
    "tts": {
      "tag": "tts_edge",
      "args": {
        "voice_name": "zh-CN-YunjianNeural",
        "language": "zh",
        "gender": "Male"
      }
    }
  },
  "config_list": []
}
```

run livekit kimi vision bot
```shell
PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True LIVEKIT_URL=$LIVEKIT_URL LIVEKIT_API_KEY=$LIVEKIT_API_KEY LIVEKIT_API_SECRET=$LIVEKIT_API_SECRET \
  python -m src.cmd.bots.main -f config/bots/livekit_describe_transformers_vision_bot.json
```

---
# deploy modal
- download moonshotai/Kimi-VL-A3B-Instruct  moonshotai/Kimi-VL-A3B-Thinking models
```shell
modal run  src/download_models.py --repo-ids "moonshotai/Kimi-VL-A3B-Instruct"
modal run  src/download_models.py --repo-ids "moonshotai/Kimi-VL-A3B-Thinking"
```

- run modal serve
```shell
# moonshotai/Kimi-VL-A3B-Instruct (or Thinking) use 2xL4 gpus like deepseek-ai/deepseek-vl2-small
IMAGE_NAME=kimi IMAGE_CONCURRENT_CN=1 IMAGE_GPU=L4:2 modal serve src/fastapi_webrtc_vision_bot_serve.py

# moonshotai/Kimi-VL-A3B-Instruct (or Thinking) use L40s gpu
IMAGE_NAME=kimi IMAGE_CONCURRENT_CN=1 IMAGE_GPU=L40s modal serve src/fastapi_webrtc_vision_bot_serve.py

# moonshotai/Kimi-VL-A3B-Instruct (or Thinking) use A100 gpu
IMAGE_NAME=kimi IMAGE_CONCURRENT_CN=1 IMAGE_GPU=A100 modal serve  src/fastapi_webrtc_vision_bot_serve.py
```

- run livekit webrtc kimi vision bot with L40s gpu device
```shell
curl --location 'https://weedge-achatbot--fastapi-webrtc-vision-kimi-bot-srv-app-dev.modal.run/bot_join/chat-room/LivekitDescribeVisionBot' \
--header 'Authorization: Bearer xxx' \
--header 'Content-Type: application/json' \
--data '{
    "chat_bot_name": "LivekitDescribeVisionBot",
    "room_name": "chat-room",
    "room_url": "",
    "token": "",
    "room_manager": {
        "tag": "livekit_room",
        "args": {
            "bot_name": "LivekitDescribeVisionBot",
            "is_common_session": false
        }
    },
    "services": {
        "pipeline": "achatbot",
        "vad": "silero",
        "asr": "sense_voice",
        "llm": "llm_transformers_manual_vision_kimi",
        "tts": "edge"
    },
    "config": {
        "vad": {
            "tag": "silero_vad_analyzer",
            "args": {
                "stop_secs": 0.7
            }
        },
        "asr": {
            "tag": "sense_voice_asr",
            "args": {
                "language": "zn",
                "model_name_or_path": "/root/.achatbot/models/FunAudioLLM/SenseVoiceSmall"
            }
        },
        "llm": {
            "tag": "llm_transformers_manual_vision_kimi",
            "args": {
                "lm_device": "cuda",
                "lm_model_name_or_path": "/root/.achatbot/models/moonshotai/Kimi-VL-A3B-Instruct",
                "chat_history_size": 0,
               "lm_attn_impl":"flash_attention_2",
                "lm_torch_dtype": "bfloat16",
                "model_type": "chat_completion"
            },
            "language": "zh"
        },
        "tts": {
            "tag": "tts_edge",
            "args": {
                "voice_name": "zh-CN-YunjianNeural",
                "language": "zh",
                "gender": "Male"
            }
        }
    },
    "config_list": []
}'
```
L40s:

image

image

- run livekit webrtc kimi vision bot with 2xL4 gpus device (use default split model)
```shell
curl --location 'https://weedge-achatbot--fastapi-webrtc-vision-kimi-bot-srv-app-dev.modal.run/bot_join/chat-room/LivekitDescribeVisionBot' \
--header 'Authorization: Bearer xxx' \
--header 'Content-Type: application/json' \
--data '{
    "chat_bot_name": "LivekitDescribeVisionBot",
    "room_name": "chat-room",
    "room_url": "",
    "token": "",
    "room_manager": {
        "tag": "livekit_room",
        "args": {
            "bot_name": "LivekitDescribeVisionBot",
            "is_common_session": false
        }
    },
    "services": {
        "pipeline": "achatbot",
        "vad": "silero",
        "asr": "sense_voice",
        "llm": "llm_transformers_manual_vision_kimi",
        "tts": "edge"
    },
    "config": {
        "vad": {
            "tag": "silero_vad_analyzer",
            "args": {
                "stop_secs": 0.7
            }
        },
        "asr": {
            "tag": "sense_voice_asr",
            "args": {
                "language": "zn",
                "model_name_or_path": "/root/.achatbot/models/FunAudioLLM/SenseVoiceSmall"
            }
        },
        "llm": {
            "tag": "llm_transformers_manual_vision_kimi",
            "args": {
                "lm_device_map":{},
                "lm_model_name_or_path": "/root/.achatbot/models/moonshotai/Kimi-VL-A3B-Instruct",
                "chat_history_size": 0,
               "lm_attn_impl":"flash_attention_2",
                "lm_torch_dtype": "bfloat16",
                "model_type": "chat_completion"
            },
            "language": "zh"
        },
        "tts": {
            "tag": "tts_edge",
            "args": {
                "voice_name": "zh-CN-YunjianNeural",
                "language": "zh",
                "gender": "Male"
            }
        }
    },
    "config_list": []
}'
```
2xL4:

image

image

---

image

---

# reference
- https://github.com/MoonshotAI/Kimi-VL
- [Kimi k1.5: Scaling Reinforcement Learning with LLMs](https://arxiv.org/abs/2501.12599) 
- SigLIP-SO-400M: [Sigmoid Loss for Language Image Pre-Training](https://arxiv.org/abs/2303.15343) | [paper code](https://github.com/google-research/big_vision)

### CODE REVIEW

**Comment by weedge** [comment]
moonshotai/Kimi-VL-A3B-Instruct: 16407.656048 M parameters
```
KimiVLForConditionalGeneration(
  (vision_tower): MoonVitPretrainedModel(
    (patch_embed): MoonVisionPatchEmbed(
      (proj): Conv2d(3, 1152, kernel_size=(14, 14), stride=(14, 14))
      (pos_emb): Learnable2DInterpPosEmb()
    )
    (encoder): MoonVitEncoder(
      (rope_2d): Rope2DPosEmb(dim=72, max_height=512, max_width=512, theta_base=10000)
      (blocks): ModuleList(
        (0-26): 27 x MoonVitEncoderLayer(
          (norm0): LayerNorm((1152,), eps=1e-05, elementwise_affine=True)
          (norm1): LayerNorm((1152,), eps=1e-05, elementwise_affine=True)
          (mlp): MLP2(
            (fc0): Linear(in_features=1152, out_features=4304, bias=True)
            (fc1): Linear(in_features=4304, out_features=1152, bias=True)
            (activation): PytorchGELUTanh()
          )
          (wqkv): Linear(in_features=1152, out_features=3456, bias=True)
          (wo): Linear(in_features=1152, out_features=1152, bias=True)
        )
      )
      (final_layernorm): LayerNorm((1152,), eps=1e-05, elementwise_affine=True)
    )
  )
  (multi_modal_projector): KimiVLMultiModalProjector(
    (pre_norm): LayerNorm((1152,), eps=1e-05, elementwise_affine=True)
    (linear_1): Linear(in_features=4608, out_features=4608, bias=True)
    (act): GELUActivation()
    (linear_2): Linear(in_features=4608, out_features=2048, bias=True)
  )
  (language_model): DeepseekV3ForCausalLM(
    (model): DeepseekV3Model(
      (embed_tokens): Embedding(163840, 2048, padding_idx=163839)
      (layers): ModuleList(
        (0): DeepseekV3DecoderLayer(
          (self_attn): DeepseekV3Attention(
            (q_proj): Linear(in_features=2048, out_features=3072, bias=False)
            (kv_a_proj_with_mqa): Linear(in_features=2048, out_features=576, bias=False)
            (kv_a_layernorm): DeepseekV3RMSNorm()
            (kv_b_proj): Linear(in_features=512, out_features=4096, bias=False)
            (o_proj): Linear(in_features=2048, out_features=2048, bias=False)
            (rotary_emb): DeepseekV3RotaryEmbedding()
          )
          (mlp): DeepseekV3MLP(
            (gate_proj): Linear(in_features=2048, out_features=11264, bias=False)
            (up_proj): Linear(in_features=2048, out_features=11264, bias=False)
            (down_proj): Linear(in_features=11264, out_features=2048, bias=False)
            (act_fn): SiLU()
          )
          (input_layernorm): DeepseekV3RMSNorm()
          (post_attention_layernorm): DeepseekV3RMSNorm()
        )
        (1-26): 26 x DeepseekV3DecoderLayer(
          (self_attn): DeepseekV3Attention(
            (q_proj): Linear(in_features=2048, out_features=3072, bias=False)
            (kv_a_proj_with_mqa): Linear(in_features=2048, out_features=576, bias=False)
            (kv_a_layernorm): DeepseekV3RMSNorm()
            (kv_b_proj): Linear(in_features=512, out_features=4096, bias=False)
            (o_proj): Linear(in_features=2048, out_features=2048, bias=False)
            (rotary_emb): DeepseekV3RotaryEmbedding()
          )
          (mlp): DeepseekV3MoE(
            (experts): ModuleList(
              (0-63): 64 x DeepseekV3MLP(
                (gate_proj): Linear(in_features=2048, out_features=1408, bias=False)
                (up_proj): Linear(in_features=2048, out_features=1408, bias=False)
                (down_proj): Linear(in_features=1408, out_features=2048, bias=False)
                (act_fn): SiLU()
              )
            )
            (gate): MoEGate()
            (shared_experts): DeepseekV3MLP(
              (gate_proj): Linear(in_features=2048, out_features=2816, bias=False)
              (up_proj): Linear(in_features=2048, out_features=2816, bias=False)
              (down_proj): Linear(in_features=2816, out_features=2048, bias=False)
              (act_fn): SiLU()
            )
          )
          (input_layernorm): DeepseekV3RMSNorm()
          (post_attention_layernorm): DeepseekV3RMSNorm()
        )
      )
      (norm): DeepseekV3RMSNorm()
    )
    (lm_head): Linear(in_features=2048, out_features=163840, bias=False)
  )
)
```
